doc/api/esm.md

```diff
@@ -1154,6 +1154,40 @@ export async function transformSource(source,
 }
 ```
 
+#### <code>getGlobalPreloadCode</code> hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+Sometimes it can be necessary to run some code inside of the same global scope
+that the application will run in. This hook allows to return a string that will
+be ran as sloppy-mode script on startup.
+
+Similar to how CommonJS wrappers work, the last statement in the code may
+evaluate to a function. If it does, it will be called immediately. The only
+argument is a `require`-like function that can be used to load builtins.
+
+If the code needs more advanced `require` features, it will have to construct
+its own `require` using  `module.createRequire()`.
+
+```js
+/**
+ * @returns {string} Code to run before application startup
+ */
+export function getGlobalPreloadCode() {
```

Comment:
Should this follow the pattern `context, defaultGetGlobalPreloadCode` that the other hooks do? So that we can add chaining of hooks at a later point.

Reply:
Good question. I did consider at some point having a `defaultGetGlobalPreloadCode`. The problem is that combining strings of code is a lot harder to get right than what most other hooks do. My gut feeling for global preload is "this is code the loader really is certain needs to run, exactly like it is". It's not like `resolve` where we actually *need* one value at the end.

So my current assumption is that if there would be multiple preload hooks, node would collect them and execute them in order. They wouldn't get to mess with each other's code.

Reply:
I think that’s what everyone here agrees may be best. E.g.:

> So my current assumption is that if there would be multiple preload hooks, node would collect them and execute them in order.

But as Bradley said, loaders shouldn’t be able to mess with each other’s code here. So a parent function wouldn’t be meaningful most likely.

Reply:
Agreed we can consider some hooks as chaining to a single result, like resolve and translate, and others running with serial execution consequences like this one.